FEATURE: support zookeeper dynamic reconfiguration

SuhwanJang opened the pull request:
**zk dynamic reconfiguration 반영** (https://github.com/jam2in/arcus-works/issues/277)

- zoo.cfg backward compatibility 지원은 별도 PR 에서 처리할 예정입니다.
- zk_reconfig 설정 명령 제공은 별도 PR에서 처리할 예정입니다.

closed PR 참고 : https://github.com/naver/arcus-memcached/pull/571

SuhwanJang commented:
@jhpark816 리뷰 반영하였습니다. closed PR : #571

이전 PR에서 추가로 수정한 사항입니다.
- config version hex 표시 
- zoo_set_servers() 수행 전 500~1000 usec random sleep
- arcus_read_ZK_config() 에서 ZNONODE handling 
- zoo_set_servers() 수행 조건 version != 0 추가

jhpark816 commented on arcus_zk.c:
```diff
@@ -1130,6 +1204,45 @@ static int arcus_parse_server_mapping(char *znode)
+        if (sm_info.zkconfig_host_buffer == NULL) {
+            arcus_conf.logger->log(EXTENSION_LOG_WARNING, NULL,
+                    "Failed to allocate zkconfig host buffer\n");
+            free(sm_info.zkconfig_data_buffer);
```
`sm_info.zkconfig_data_buffer = NULL;` 설정해 두는 것이 안전할 것 같습니다.

jhpark816 reviewed (commented):
리뷰 완료.

config znode가 없는 경우를 포함하여, reconfig 관련 플래그는 아래 중의 하나가 됩니다.
1. zk_reconfig = true & update_zkconfig = true & NO watching 상태 : 정상
2. zk_reconfig = true & update_zkconfig = false & watching 상태: 정상
3. zk_reconfig = true & update_zkconfig = false & NO watching 상태: NOZNODE로 중단
4. zk_reconfig = false : disabled

향후, ZK reconfig 기능에 대해 아래 명령을 제공되어야 합니다. 참고 바라빈다.
- enable/disable 시키는 명령
   - 위의 3번과 4번에 대해 모두 enable하는 기능
   - 위의 1번과 2번에 대해 모두 disable하는 기능 : No watching 상태가 되어야 함.
- 상태를 보여주는 명령

jhpark816 commented:
NOZNODE 오류의 처리가 맞는 지는 다시 의견을 들었으면 합니다.
- 현재처럼 watching 중단이 하나의 방안이 되고
- 긴 주기(예, 1 or 10분)로 retry하는 것도 하나의 방안이 될 수 있을 것 같습니다.

SuhwanJang commented on arcus_zk.c:
```diff
@@ -1381,6 +1503,139 @@ static int sm_reload_cache_list_znode(zhandle_t *zh, bool *retry)
+
+        if (sm_info.zkconfig_version == -1) {
+            sm_info.zkconfig_version = version;
+        } else if (version != 0 && version > sm_info.zkconfig_version) {
```
zkconfig_version 이 0인 상태에서 reconfig 요청을 받으면 zoo_set_servers()를 수행하지 않게 됩니다. 따라서 zkconfig_version 초기값을 0으로 하면 안될 것 같습니다.
- ensemble member 3개 구성
   - 1 ZK (F, reconfigEnabled = false) / 2 ZK (L, reconfigEnabled = false) / 3 ZK (F, reconfigEnabled = true)
- 구동 후 config version 는 다음과 같음. 
   - 1 ZK : 0 / 2 ZK : 0 / 3 ZK : 100000000
- kill Leader 후 3이 Leader 가 됨. 이후 config version
   - 1 ZK(F) : 0 / 3 ZK(L) : 100000000
-  2 restart
   - 1 ZK(F) : 0 / 2 ZK(F) : 0 / 3 ZK(L) : 100000000
-  reconfig 수행
   - 1 ZK(F) : 200000009 / 2 ZK(F) : 200000009 / 3 ZK(L) : 200000009

jhpark816 replied:
초기 값이 0이면 안되겠네요.
version 값이 16진수이므로, unsinged integer 값을 가질 것입니다.
초기 값으로 -1을 사용해도 되는 지가 궁금해집니다.

SuhwanJang replied:
ZK server 는 내부적으로 config version 을 long type 으로 쓰고 있습니다. JAVA long 은 C long long or int64_t 와 동일하므로 int64_t 쓰도록 수정하겠습니다.
```
ReconfigCommand.java
     /* version - version of config from which we want to reconfigure - if current config is different
      * reconfiguration will fail. Should be committed from the CLI to disable this option.
      */
     long version = -1;
```

SuhwanJang commented:
> NOZNODE 오류의 처리가 맞는 지는 다시 의견을 들었으면 합니다.
> 
> * 현재처럼 watching 중단이 하나의 방안이 되고
> * 긴 주기(예, 1 or 10분)로 retry하는 것도 하나의 방안이 될 수 있을 것 같습니다.

znode가 없어 오류가 나온 것이므로 긴 주기로 retry 하는 것은 의미가 없어 보입니다.

SuhwanJang commented on arcus_zk.c:
```diff
@@ -1381,6 +1503,139 @@ static int sm_reload_cache_list_znode(zhandle_t *zh, bool *retry)
+                arcus_conf.logger->log(EXTENSION_LOG_WARNING, NULL,
+                    "Failed to update ZK servers. zoo_set_servers() failed: %s\n", zerror(rc));
+            }
+            sm_info.zkconfig_version = version;
```
공식 가이드 예제에서는 zkconfig_version 세팅은 zoo_set_servers 성공 관계없이 수행하고 있습니다.
아마 zoo_set_servers는 client 에서의 처리가 실패한 것이므로 config version 은 ZK server의 config version & 다른 ZK client 의 config version 과 동일한 값을 갖도록 하려는 의도인 것 같습니다.
```
public void processResult(int rc, String path, Object ctx, byte[] data, Stat stat) {
    if (path!=null &&  path.equals(ZooDefs.CONFIG_NODE)) {
        String config[] = ConfigUtils.getClientConfigStr(new String(data)).split(" ");   // similar to config -c
        long version = Long.parseLong(config[0], 16);
        if (this.configVersion == null){
             this.configVersion = version;
        } else if (version > this.configVersion) {
            hostList = config[1];
            try {
                // the following command is not blocking but may cause the client to close the socket and
                // migrate to a different server. In practice its better to wait a short period of time, chosen
                // randomly, so that different clients migrate at different times
                zk.updateServerList(hostList);
            } catch (IOException e) {
                System.err.println("Error updating server list");
                e.printStackTrace();
            }
            this.configVersion = version;
} } }
```

의견주신 것처럼 성공한 경우에만 version 설정이 의미상 맞아보이는데, 언제 설정하던 동작하는데 문제는 없어보입니다. 
최종 의견 한번만 더 주시면 반영하겠습니다.

jhpark816 replied:
@SuhwanJang 
생각해 보니, 
- zoo_set_servers()가 실패하는 경우는 어떤 리소스 부족한 경우일 거라 생각되고,
  이 경우, 계속 retry 하더라도 동일 오류가 반복하여 발생할 거라 생각됩니다.
- 결국, 오류가 발생하면, 운영자가 detect하고 조치해야 할 사항으로 보이네요.

반복된 오류 출력을 막기 위하여, 가이드 문서대로 따르도록 하시죠.

SuhwanJang commented:
@jhpark816 리뷰 반영하였습니다.
- zkconfig_version type 수정 : long => int64_t
- sm_info.zkconfig_data_buffer = NULL 설정
- zkconfig_version 초기값 : 0
- zkconfig_version != 0 인 경우에 zoo_set_servers() 수정
- random sleep : 0 ~ 1000 usec.
- sm_reload_ZK_config function return type 수정 : int => void.

SuhwanJang commented:
@jhpark816 memcached test log
// memcached 구동 [ensemble : 0.0.0.0:2150,0.0.0.0:2151,0.0.0.0:2152,0.0.0.0:2153]

```
...
2021-03-04 12:24:13,792:42996(0x7f4e93598700):ZOO_INFO@check_events@2500: initiated connection to server [10.34.92.23
4:2150]
2021-03-04 12:24:13,796:42996(0x7f4e93598700):ZOO_INFO@check_events@2552: session establishment complete on server [1
0.34.92.234:2150], sessionId=0x101aa7d3c550012, negotiated timeout=30000
ZooKeeper client initialized. (ZK session timeout=30 sec)
Found the valid server mapping. service_code=test
ZooKeeper dynamic reconfiguration is enabled.
Joined Arcus cloud at /arcus/cache_list/test/10.34.92.234:11300-jam2in-t002(ver=0)
Memcached joined Arcus cache cloud for "test" service (took 18843 microsec)
update cluster config...
server[0] = 10.34.92.234:11300-jam2in-t002
Memcached is watching Arcus cache cloud for "/arcus/cache_list/test"
Heartbeat thread is running.
update cluster config...
server[0] = 10.34.92.234:11300-jam2in-t002
```

//reconfig remove  (port:2153)

```
EVENT from ZK config: state=3, path=/zookeeper/config
Updated ZK servers... ZK servers[0.0.0.0:2150,0.0.0.0:2151,0.0.0.0:2152], version[100000065]
```

//reconfig add (port:2153)

```
EVENT from ZK config: state=3, path=/zookeeper/config
Updated ZK servers... ZK servers[0.0.0.0:2150,0.0.0.0:2151,0.0.0.0:2152,0.0.0.0:2153], version[100000068]
```

jhpark816 commented on arcus_zk.c:
```diff
@@ -1260,6 +1374,15 @@ static int arcus_create_ephemeral_znode(zhandle_t *zh)
 {
+#ifdef ENABLE_ZK_RECONFIG
+    /* check zookeeper dynamic reconfiguration support */
+    if (arcus_check_zoo_getconfig_supported(main_zk->zh) != 0) {
```
arcus_register_cache_instance() 안에서 호출한다면, `main_zk->zh`가 아니라 `zh`이어야 합니다.

jhpark816 replied:
arcus_register_cache_instance() 안에서 호출하는 것이 적당하지 않습니다.
sm_init() 수행 직전에 호출하는 것이 적당합니다.

jhpark816 commented on arcus_zk.c:
```diff
@@ -219,6 +232,10 @@ static void arcus_zk_watcher(zhandle_t *wzh, int type, int state,
+#ifdef ENABLE_ZK_RECONFIG
+static void arcus_zkconfig_watcher(zhandle_t *zh, int type, int state,
+                                   const char *path, void *ctx);
+#endif
```
zkreconfig 관련 코드가 cache_list 관련 코드 위에 있는 것이 좋겠습니다.
따라서, arcus_cache_list_watcher() 위로 옮기시죠.

jhpark816 commented on arcus_zk.c:
```diff
@@ -233,6 +250,9 @@ int mc_hb(void *context);     // memcached self-heartbeat
     bool update_cache_list;
+#ifdef ENABLE_ZK_RECONFIG
+    bool update_zkconfig;
+#endif
```
여기도 update_cache_list 위로 이동

jhpark816 commented on arcus_zk.c:
```diff
@@ -505,6 +536,33 @@ arcus_cache_list_watcher(zhandle_t *zh, int type, int state, const char *path, v
+    sm_wakeup(true);
+    sm_unlock();
+}
+#endif
```
여기도 arcus_cache_list_watcher() 위로 이동

jhpark816 commented on arcus_zk.c:
```diff
@@ -1536,6 +1790,11 @@ void arcus_zk_init(char *ensemble_list, int zk_to,
+    if (arcus_conf.zk_reconfig) {
+        sm_info.request.update_zkconfig = true;
+    }
+#endif
```
여기도 update_cache_list 위로 이동

jhpark816 commented on arcus_zk.c:
```diff
@@ -1763,6 +2032,11 @@ int arcus_zk_rejoin_ensemble()
+        if (arcus_conf.zk_reconfig) {
+            sm_info.request.update_zkconfig = true;
+        }
+#endif
```
여기도 update_cache_list 설정 위로 이동

jhpark816 commented on arcus_zk.c:
```diff
@@ -519,6 +577,22 @@ arcus_read_ZK_children(zhandle_t *zh, const char *zpath, watcher_fn watcher,
+    }
+    return 1;
+}
+#endif
```
이 코드도 arcus_read_ZK_children() 위로 이동

jhpark816 reviewed (commented):
review 완료

jhpark816 commented:
추가 사항으로, 아래 수정 사항들 중
- zkconfig_version 초기 값은 -1이고,
- sm_reload_ZK_config() 리턴 타입을 int => void로 수정한 이유는 무엇인가요 ?

---------------
- zkconfig_version type 수정 : long => int64_t
- sm_info.zkconfig_data_buffer = NULL 설정
- zkconfig_version 초기값 : 0
- zkconfig_version != 0 인 경우에 zoo_set_servers() 수정
- random sleep : 0 ~ 1000 usec.
- sm_reload_ZK_config function return type 수정 : int => void.

SuhwanJang commented:
@jhpark816 리뷰 반영하였습니다. 
> 추가 사항으로, 아래 수정 사항들 중
> 
> * zkconfig_version 초기 값은 -1이고,
> * sm_reload_ZK_config() 리턴 타입을 int => void로 수정한 이유는 무엇인가요 ?
> 

sm_reload_ZK_config() 리턴 타입을 확인하는 부분이 sm_state_thread() 한 군데 있었습니다. (아래 기존 코드 참고)
기존에 리턴 값은 0 아니면 1 였는데, shutdown_by_me = true 조건을 충족하는 경우가 없기 때문에 의미가 없어보여 void로 수정하였습니다.

```
 #ifdef ENABLE_ZK_RECONFIG
         if (smreq.update_zkconfig) {
             if (sm_reload_ZK_config(main_zk->zh, &sm_retry) < 0) {
                 shutdown_by_me = true; break;
             }
             if (arcus_zk_shutdown)
                 break;
         }
 #endif
```

- arcus_check_zoo_getconfig_supported() 호출은 sm_init() 에서 sm_info struct initialize 를 수행하기 때문에 sm_init() 이후에 하도록 수정하였습니다.
